# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1366145800 14400
#      Tue Apr 16 16:56:40 2013 -0400
# Node ID 8701792e16ecbd2c12dc2895c48373ee74672346
# Parent  01541f7321f78fe67870c28972c72607324d70ce
include column in workspace viewer for storage class

* workspace-model.cc (workspace_model::workspace_model): New column in
for storage class.
(workspace_model::data): Set values for storage class column.
(symbol_table::do_workspace_info): Also insert a tag in storage for
automatic, formal, hidden, and inherited storage classes.

diff --git a/libgui/src/workspace-model.cc b/libgui/src/workspace-model.cc
--- a/libgui/src/workspace-model.cc
+++ b/libgui/src/workspace-model.cc
@@ -33,16 +33,17 @@ along with Octave; see the file COPYING.
 
 workspace_model::workspace_model (QObject *p)
   : QAbstractTableModel (p)
 {
   _columnNames.append (tr ("Name"));
   _columnNames.append (tr ("Class"));
   _columnNames.append (tr ("Dimension"));
   _columnNames.append (tr ("Value"));
+  _columnNames.append (tr ("Storage Class"));
 }
 
 int
 workspace_model::rowCount(const QModelIndex& p) const
 {
   return _symbols.size ();
 }
 
@@ -104,16 +105,34 @@ workspace_model::data (const QModelIndex
         case 2:
           retval = QVariant (_dimensions[idx.row()]);
           break;
 
         case 3:
           retval = QVariant (_values[idx.row()]);
           break;
 
+        case 4:
+          {
+            QChar c = _scopes[idx.row()];
+
+            if (c == 'g')
+              retval = QVariant (tr ("global"));
+            else if (c == 'p')
+              retval = QVariant (tr ("persistent"));
+            else if (c == 'a')
+              retval = QVariant (tr ("automatic"));
+            else if (c == 'f')
+              retval = QVariant (tr ("function parameter"));
+            else if (c == 'h')
+              retval = QVariant (tr ("hidden"));
+            else if (c == 'i')
+              retval = QVariant (tr ("inherited"));
+          }
+
         default:
           break;
         }
     }
 
   return retval;
 }
 
diff --git a/libinterp/interpfcn/symtab.cc b/libinterp/interpfcn/symtab.cc
--- a/libinterp/interpfcn/symtab.cc
+++ b/libinterp/interpfcn/symtab.cc
@@ -1398,16 +1398,24 @@ symbol_table::do_workspace_info (void) c
             {
               dim_vector dv = val.dims ();
 
               char storage = ' ';
               if (sr.is_global ())
                 storage = 'g';
               else if (sr.is_persistent ())
                 storage = 'p';
+              else if (sr.is_automatic ())
+                storage = 'a';
+              else if (sr.is_formal ())
+                storage = 'f';
+              else if (sr.is_hidden ())
+                storage = 'h';
+              else if (sr.is_inherited ())
+                storage = 'i';
 
               workspace_element elt (storage, nm, val.class_name (),
                                      val.short_disp (), dv.str ());
 
               retval.push_back (elt);
             }
         }
     }
